File: FlowCyPy/fluidics/population_events.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from dataclasses import dataclass, field
from typing import Any

from FlowCyPy.sub_frames.events import EventDataFrame


@dataclass
class PopulationEvents:
    """
    Structured container storing all event level data for one population.

    This class is the canonical internal event representation used by the
    simulation pipeline. It separates three categories of information:

    1. Event level tabular data
       Stored in ``dataframe``. These are per event quantities such as time,
       position, velocity, and detector specific amplitudes.

    2. Semantic simulation objects
       Stored explicitly in ``population`` and ``sampling_method``.

    3. Population level metadata
       Stored in ``metadata``. These are scalar or aggregate quantities such as
       mean velocity, expected occupancy per time bin, or diagnostic traces.

    Parameters
    ----------
    dataframe : EventDataFrame
        Event level table.
    population : object
        Population object that generated the events.
    sampling_method : object
        Sampling method used to generate the event block.
    name : str
        Population name.
    scatterer_type : str
        Scatterer type label associated with the population.
    metadata : dict[str, Any], optional
        Additional population level metadata.
    """

    dataframe: EventDataFrame
    population: object
    sampling_method: object
    name: str
    scatterer_type: str
    metadata: dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        self.dataframe.scatterer_type = self.scatterer_type

        if "units" not in self.dataframe.attrs:
            self.dataframe.attrs["units"] = {}

    def __len__(self) -> int:
        """
        Return the number of stored events.

        Returns
        -------
        int
            Number of rows in the underlying dataframe.
        """
        return len(self.dataframe)

    def __getitem__(self, key: str):
        """
        Return one dataframe column.

        Parameters
        ----------
        key : str
            Column name.

        Returns
        -------
        Any
            Requested column. If the column has a registered unit, a quantity is returned.
        """
        return self.dataframe[key]

    def __setitem__(self, key: str, value: Any) -> None:
        """
        Assign one dataframe column.

        Parameters
        ----------
        key : str
            Column name.
        value : Any
            Values to assign.
        """
        self.dataframe[key] = value

    @property
    def empty(self) -> bool:
        """
        Whether this event block is empty.

        Returns
        -------
        bool
            ``True`` if the dataframe has no rows.
        """
        return self.dataframe.empty

    @property
    def columns(self):
        """
        Return dataframe columns.

        Returns
        -------
        pandas.Index
            Column index.
        """
        return self.dataframe.columns

    def get_quantity(self, column_name: str):
        """
        Return the specified column as a quantity.

        Parameters
        ----------
        column_name : str
            Target column name.

        Returns
        -------
        pint.Quantity or numpy.ndarray
            Column values with units if the column is unit aware.
        """
        return self.dataframe.get_quantity(column_name)

    def set_quantity_column(self, column_name: str, value) -> None:
        """
        Store a quantity column in the underlying unit aware dataframe.

        Parameters
        ----------
        column_name : str
            Name of the target column.
        value : pint.Quantity
            Quantity to store.
        """
        self.dataframe.set_column(column_name=column_name, values=value)

    def to_dataframe(self, include_metadata_in_attrs: bool = True) -> EventDataFrame:
        """
        Export this event block as a standalone dataframe.

        Parameters
        ----------
        include_metadata_in_attrs : bool, default=True
            Whether to copy metadata into ``DataFrame.attrs`` on export.

        Returns
        -------
        EventDataFrame
            Exported dataframe copy.
        """
        dataframe = self.dataframe.copy(deep=True)
        dataframe.attrs = {
            key: value.copy() if isinstance(value, dict) else value
            for key, value in self.dataframe.attrs.items()
        }

        dataframe.scatterer_type = self.scatterer_type
        dataframe.attrs["Name"] = self.name
        dataframe.attrs["PopulationType"] = self.scatterer_type
        dataframe.attrs["ScattererType"] = self.scatterer_type
        dataframe.attrs["SamplingMethod"] = self.sampling_method.__class__.__name__

        if include_metadata_in_attrs:
            for key, value in self.metadata.items():
                dataframe.attrs[key] = value

        return dataframe

File: FlowCyPy/cpp/fluidics/population_events/population_events.cpp
#include "population_events.h"

#include <stdexcept>

namespace {

py::dict copy_mapping_values(const py::dict& source) {
    py::dict output;
    for (const auto& item : source) {
        py::object value = py::reinterpret_borrow<py::object>(item.second);
        output[item.first] = py::isinstance<py::dict>(value)
            ? value.attr("copy")()
            : value;
    }
    return output;
}

py::object dataframe_copy(const py::object& dataframe) {
    py::kwargs keywords;
    keywords["deep"] = true;
    return dataframe.attr("copy")(**keywords);
}

void set_dataframe_column(
    const py::object& dataframe,
    const std::string& column_name,
    const py::object& values
) {
    py::kwargs keywords;
    keywords["column_name"] = column_name;
    keywords["values"] = values;
    dataframe.attr("set_column")(**keywords);
}

}

PopulationEvents::PopulationEvents(
    py::object dataframe,
    py::object population,
    py::object sampling_method,
    std::string name,
    std::string scatterer_type,
    py::object metadata
)
    : dataframe(std::move(dataframe)),
      population(std::move(population)),
      sampling_method(std::move(sampling_method)),
      name(std::move(name)),
      scatterer_type(std::move(scatterer_type)),
      metadata(metadata.is_none() ? py::dict() : metadata.cast<py::dict>()) {
    synchronize_dataframe_metadata();
}

void PopulationEvents::synchronize_dataframe_metadata() {
    dataframe.attr("scatterer_type") = scatterer_type;
    py::dict attrs = dataframe.attr("attrs").cast<py::dict>();
    if (!attrs.contains("units")) {
        attrs["units"] = py::dict();
    }
}

std::size_t PopulationEvents::size() const {
    return py::len(dataframe);
}

std::shared_ptr<PopulationEvents> PopulationEvents::copy() const {
    py::object copied_dataframe = dataframe_copy(dataframe);
    copied_dataframe.attr("attrs") = copy_mapping_values(
        dataframe.attr("attrs").cast<py::dict>()
    );

    return std::make_shared<PopulationEvents>(
        copied_dataframe,
        population,
        sampling_method,
        name,
        scatterer_type,
        copy_mapping_values(metadata)
    );
}

py::object PopulationEvents::get_item(const py::object& key) const {
    return dataframe.attr("__getitem__")(key);
}

void PopulationEvents::set_item(
    const py::object& key,
    const py::object& value
) {
    dataframe.attr("__setitem__")(key, value);
}

bool PopulationEvents::empty() const {
    return dataframe.attr("empty").cast<bool>();
}

py::object PopulationEvents::columns() const {
    return dataframe.attr("columns");
}

py::object PopulationEvents::get_quantity(
    const std::string& column_name
) const {
    return dataframe.attr("get_quantity")(column_name);
}

void PopulationEvents::set_quantity_column(
    const std::string& column_name,
    const py::object& value
) {
    set_dataframe_column(dataframe, column_name, value);
}

py::object PopulationEvents::to_dataframe(
    const bool include_metadata_in_attrs
) const {
    py::object output = dataframe_copy(dataframe);
    output.attr("attrs") = copy_mapping_values(
        dataframe.attr("attrs").cast<py::dict>()
    );
    output.attr("scatterer_type") = scatterer_type;

    py::dict attrs = output.attr("attrs").cast<py::dict>();
    attrs["Name"] = name;
    attrs["PopulationType"] = scatterer_type;
    attrs["ScattererType"] = scatterer_type;
    attrs["SamplingMethod"] = sampling_method.attr("__class__").attr("__name__");

    if (include_metadata_in_attrs) {
        for (const auto& item : metadata) {
            attrs[item.first] = item.second;
        }
    }

    return output;
}
